ci/build-channel/src/main.rs

```diff
@@ -152,22 +118,68 @@ fn validate_components(channel: &str, components: &HashMap<String, Version>) ->
+fn write_nightly_document(document: &mut Document, components: Vec<Component>) -> Result<()> {
+    let mut data = Vec::new();
+    let nightly_release_url = format!(
+        "https://api.github.com/repos/FuelLabs/sway-nightly-binaries/releases/tags/nightly-{}",
+        *TODAY
+    );
 
-    for package in args.packages {
-        component_versions.insert(package.0, package.1);
+    let resp = ureq::get(&nightly_release_url).call()?;
+    resp.into_reader().read_to_end(&mut data)?;
+    let release: Release = serde_json::from_str(&String::from_utf8_lossy(&data))?;
+
+    for asset in release.assets {
+        for component in &components {
+            // Asset name example: fuel-core-0.15.1+nightly.20230111.a5514420e5-x86_64-unknown-linux-gnu.tar.gz
+            // If an asset's name matches a component's declared tarball_prefix in components.toml,
+            // we want to store the download information in a channel.
+            if let Some(stripped) = asset.name.strip_prefix(&component.tarball_prefix) {
+                println!("\nWriting package info for component '{}'", &component.name);
+                document["pkg"][&component.name] = implicit_table();
+                document["pkg"][&component.name]["target"] = implicit_table();
+
+                // Example output: Some((0.15.1+nightly.20230111.a5514420e5, x86_64-unknown-linux-gnu.tar.gz))
+                // We want to record the version and target in the channel toml.
+                let split = stripped[1..].split_once('-');
+                if let Some((version, tarball_name)) = split {
+                    document["pkg"][&component.name]["version"] = value(version.to_string());
+
+                    // Example output: Some((x86_64-unknown-linux-gnu, tar.gz))
+                    if let Some((target, _)) = tarball_name.split_once('.') {
+                        let mut data = Vec::new();
+
+                        document["pkg"][&component.name]["target"][target.to_string()] =
+                            implicit_table();
+                        document["pkg"][&component.name]["target"][target.to_string()]["url"] =
+                            value(&asset.browser_download_url);
+
+                        if let Ok(res) = ureq::get(&asset.browser_download_url).call() {
+                            res.into_reader().read_to_end(&mut data)?;
+                            let mut hasher = Sha256::new();
+                            hasher.update(data);
+                            let actual_hash = format!("{:x}", hasher.finalize());
+
+                            println!(
+                                "url: {}\nhash: {}",
+                                &asset.browser_download_url, &actual_hash
+                            );
+                            document["pkg"][&component.name]["target"][target.to_string()]
+                                ["hash"] = value(actual_hash);
+                        };
+                    };
+                }
```

Comment:
Rather than writing directly to the toml document, one alternative approach that might be a bit more type-safe is to create a Rust type to represent the data you want, populate that, then serialize that type to `toml` afterwards. This helps to avoid typos in strings, and you get the side-benefit of being able to use the same type for deserialization.

Reply:
I'll create an issue for this for now, but this might need changing up the flow/logic because of the organization of the toml file.